sync/sync.go

```diff
@@ -61,11 +61,26 @@ func (d *MutexDatastore) Delete(key ds.Key) (err error) {
 	return d.child.Delete(key)
 }
 
-// KeyList implements Datastore.KeyList
+// Query implements Datastore.Query
 func (d *MutexDatastore) Query(q dsq.Query) (dsq.Results, error) {
 	d.RLock()
 	defer d.RUnlock()
-	return d.child.Query(q)
+	results, err := d.child.Query(q)
+	if err != nil {
+		return nil, err
+	}
+
+	// Apply the entire query while locked. Non-sync datastores may not
+	// allow concurrent queries.
+	var res []dsq.Result
+	for {
+		next, ok := results.NextSync()
```

Comment:
Why not just use `results.Rest()` and pass the resulting `[]Entries` to `ResultsWithEntries`? To pass errors exactly?

Reply:
I wanted to be able to return a bunch of results and then an error. However, in retrospect, that's probably not worth it.